### ISSUE TITLE
Zuul not forwarding client certificate to backend service

### ISSUE THREAD

**Issue by qcastel** [spring-cloud/spring-cloud-netflix#2498]
Hello!

We noticed that our zuul application is not forwarding the client certificate to the backend service.

For the moment, the closest answer I found related to that problem is a question from Andrew Serff which is unfortunately unanswered.

Our configuration of Zuul is pretty basic:

```
zuul:
  sensitiveHeaders: Set-Cookie
  routes:
    rs-api:
      path: /rs-api/**
      serviceId: rs-api

ribbon:
  eureka:
    enabled: true

eureka:
  client:
    registerWithEureka: false

server:
    hostname: XXX
    port: 8089
    ssl:
        enabled: true
        key-alias: "XXX"
        clientAuth: want
```

```
@SpringBootApplication
@EnableZuulProxy
@EnableEurekaClient
@RestController
public class ZuulServerApplication {

    public static void main(String[] args) {
        new SpringApplicationBuilder(ZuulServerApplication.class).web(true).run(args);
    }
}
```

Let me know if you need more information from me. I'm looking at how I can do that using a custom filter at the moment.

Thanks!

**Comment by qcastel** [comment]
Hello!

I though it may be easier if I reproduce it in a dummy project: 
https://github.com/qcastel/forgerockmtlswithzuul

I didn't succeeded to reproduce the issue, as I hit another zuul issue. I recon you will be able to know the problem straight aware, where me, I will spend hours trying to figure out how to fix it.

I created a readme, should help you setting up the project.

Quentin

**Comment by ryanjbaxter** [comment]
I am having a hard time understanding what you expect to happen and what the problem is.  I read the README and it sounds like if you added the certificate to the JVM it worked fine, is that correct?

**Comment by qcastel** [comment]
Hello @ryanjbaxter,

Thanks for looking at it.

If I had the certificate to the JVM cacerts, zuul is not complaining anymore that the hello world service has an untrusted certificate. I can probably open another ticket for it.

My real problem is that zuul is not sending the user certificate to the hello world service:
The hello world services has MTLS implemented and is expected to received the user certificate. It works fine if I hit the hello world directly but not through Zuul.

**Comment by qcastel** [comment]
Hello @ryanjbaxter,

I did some investigation today to work on my sample forgerockmtlswithzuul.
I managed at the end to reproduce the issue I had, but for that, I had to workaround a few zuul issues (I raised other issues to track them).

Now, the sample project should allow you to reproduce the issue:

You will see that Zuul is not passing through the client certificate to the hello world application.

**Comment by andrewserff** [comment]
@qcastel You are correct in that we never found a solution to forwarding the actual client certificate from Zuul to a back end microservice.  However, we worked around this problem by using Spring Session Replication.  You can see my [answer here](https://stackoverflow.com/questions/33921375/zuul-api-gateway-authentication/33922285#33922285) (I'm not sure if this is the one you were referencing or not). 

There isn't a standard way to forward a client cert though. It's not possible to forward the clients private key because Zuul doesn't have it. You might be able to write a Zuul filter that would take the client cert and put it in a Header like how you can with Apache (`SSLOptions +ExportCertData`). But I never tried this and your back end services would have to know to look there for that header.  

Hope that helps.

**Comment by ryanjbaxter** [comment]
@andrewserff is right.  Zuul is a proxy it cannot forward certificates.  The HTTP Client library proxying the request does not have the certificate so it cant make the secure connection and considers the server untrusted.

**Comment by qcastel** [comment]
Thanks @andrewserff and @ryanjbaxter !

I did go for the option 2: Enabling MTLS in Zuul, sending the certificate subject in the X-Client-Cert header, that then the application Hello World can read it and authenticate the original client.

I also added MTLS between Zuul and the Hello World application, to make sure to only read the X-Client-Cert if it is send by Zuul and not by another application.

I pushed the change in my demo environment, if anyone interested: https://github.com/qcastel/forgerockmtlswithzuul

Thanks again for the help!
